rep-0149.rst

```diff
@@ -346,8 +346,8 @@ Example
     </description>
     <maintainer email="someone@example.com">Someone</maintainer>
 
-    <license>BSD</license>
-    <license file="LICENSE">LGPL</license>
+    <license file="LICENSE">BSD-3-Clause</license>
```

Comment:
One suggestion I have here is to add a `type` attribute to the license tag.  The `type` attribute would describe what kind of license tag is being used, either an `SPDX` identifier, or a `freeform` identifier.  If no `type` attribute is specified, we would assume `freeform`.

Thus you could have:

* `<license>BSD</license>` - a common license tag today, which can be assumed to be a "freeform" license tag (and thus nothing can mechanically be determined about it)
* `<license type='freeform'>BSD</license>` - which is the same as above but more explicit
* `<license type='spdx'>BSD-3-Clause</license>` - which is very explicitly using the SPDX identifier, and thus can be mechanically verified

With this in place, we could put in linters (for instance) that would complain when the "type" is SPDX, but the actual text doesn't match a known SPDX identifier.  We could also have linters that would complain if any packages in the workspace don't use the "spdx" type (though we'd have to do a lot of work to enable that by default).

Does that make sense?

Reply:
Would it not be simpler to require the use of SPDX identifiers?

Other efforts to standardise licensing information (like [REUSE](https://reuse.software/spec/)) have gone that way and it makes parsing and validation much more straightforward.

Reply:
Ideally, yes, we would require SPDX identifiers.

The issue is backwards compatibility.  There are thousands of ROS (1 & 2) packages out there in the wild, and almost none of the package.xml files in them follow SPDX.  Explicitly putting a type field in will make it easier for tools (and humans) to determine whether the field is expected to be SPDX or not.

Reply:
I like the idea of introducing a type attribute. There are many ROS 1 packages that simply specify `BSD` as license identifier, which is not enough information to infer the exact license.

> which can be assumed to be a "freeform" license tag (and thus nothing can mechanically be determined about it)

I would be a bit more optimistic for the linter and in particular for generating a debian/copyright file:

  1) `<license file="path_to_license.txt">BSD</license>`: This is enough information to generate a Debian copyright file. The generator (and the linter) should check whether `path_to_license.txt` really contains a BSD license text.
  2) `<license>BSD</license>` (or `<license source-files="*">BSD</license>`) with a file `LICENSE` in the root of the repo: Again, this is enough information to generate a Debian copyright file in my opinion.
  3) `<license>Some SPDX full name or SPDX identifier</license>`: Again enough information to generate a Debian copyright file (and to validate it against a potential LICENSE file in the package or repo root).
  4) `<license file="custom_license_text.txt">MyCustomLicenseIdentifier</license>`: Also this would allow the generation of a debian/copyright file according to https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/. ("_If there are licenses present in the package without a standard short name, an arbitrary short name may be assigned for these licenses. These arbitrary names are only guaranteed to be unique within a single copyright file._")

Even the case that a debian/copyright file is generated where the license short name does not match the license text I do not consider dramatic: It simply repeats the inconsistent license names and texts from the source code and should be discovered by relevant tools applied to the debian/copyright file.

As a consequence, the `type="spdx"` attribute is not really necessary but its use would introduce a safety layer that prevents that a typo in an SPDX license identifier leads to the interpretation as a custom license identifier (i.e., as case 4).

What we must prevent in any case is that the future debian/copyright file generator refines the license specification for which there is no evidence - which is the case for some ROS 1 packages that specify `<license>BSD</license>` without any further license headers or `LICENSE` file.

Reply:
I agree with @gavanderhoorn; we should assume that all license strings are SPDX and emit an error when it's an invalid string *and* no `file` attribute is specified. Almost every possible license string is ambiguous without a `LICENSE` file, which the tooling should rightfully complain about. But we also need to let packages use custom licenses without complaining that they didn't use an SPDX identifier, so we can say the tag must either include `file` (which implies the proposed `type='freeform'`) or have a valid SPDX identifier (which implies the proposed `type='spdx'`).

@clalancette is of course right that thousands of old default-created packages will start getting errors/warnings, but in my view, that's a good thing, since an ambiguous or, worse, contradictory license means that nobody but the author can currently use those packages with confidence that the license you think you have is the license you actually have. And that's a warning we should give to people who wish to use a package in their own project: "this package's license may be unknown; proceed with caution."

Reply:
I don't feel that mandating the license type value to be SPDX is a good thing to do *in the specification*. This is the kind of decision that I believe could come back to haunt us when The Next Cool Licensing Standard comes around.

Enforcing a schema seems appropriate - maybe state that the license identifier include only alphanumerics, dots, and dashes - but forcing a specific standard here will only force folks in corner cases to put the WRONG value rather than a less-than-optimal value.

I feel strongly that if SPDX is mentioned in this REP, it be a strong recommendation and not a requirement. Additionally, if it were to be a requirement, that would constitute a breaking change and require a rev to the package.xml specifications for sure.

Recommending a specific class of license identifier is the job of a linter.

Reply:
>  if the package format version would be increased, tools could just assume "v4 manifest->spdx".

I'm trying to avoid bumping the format version.  While that is not a hard requirement, bumping the format version has implications throughout the software stack (catkin_pkg, bloom, ros-buildfarm, etc).  If we really *have* to do it, we can, but I prefer not to.

>  but don't we have the same problem with the attributes and elements proposed here? All the packages in existence today would also not have those.

All of the rest of the enhancements here are adding additional information, while changing the semantics of this field is a backwards-compatibility break.  That's why I'm trying to make SPDX opt-in.

> @clalancette is of course right that thousands of old default-created packages will start getting errors/warnings, but in my view, that's a good thing, since an ambiguous

The problem is that we'll get lots of grumpy developers, since now we are causing errors for them that hadn't been errors before.

And the thing is we can avoid all of that if we just make the SPDX identifier opt-in.  We add in the tag, allow it to be either 'freeform' or 'spdx', and then we start updating the ones we care about.  We can update the linters to have an opt-in check for SPDX licenses, and then slowly update the ecosystem to do the right thing.  We can change `ros2 pkg create` to create packages with the SPDX identifier.  We can update the documentation at https://docs.ros.org to show all examples using the SPDX identifier.  Eventually the ecosystem will follow, without making all developers unhappy at a flag day (by changing the package format 3 semantics), or causing a bunch of additional work in downstream packages (by bumping to package format 4).

Reply:
> I don't feel that mandating the license type value to be SPDX is a good thing to do in the specification. This is the kind of decision that I believe could come back to haunt us when The Next Cool Licensing Standard comes around.

I'm not sure I agree with this assessment.  While it is true that we use some REPs as specifications, we also use them as prescriptive articles, like in https://github.com/ros-infrastructure/rep/blob/master/rep-0145.rst .  That article (for instance) lays out specific frames, and even topics that IMUs should conform to.

And the thing is, if we go with my suggestion of having a "type" attribute, we can easily accommodate the next licensing standard, just by adding a new type.

So while I agree that we should have support in the linters for checking, I also think that prescribing some types here (with the default being "freeform") is OK.

Reply:
I would make this stronger and consider it a 'wasted' opportunity if we don't prescribe the use of certain license identifiers.

Too many parts of ROS are *freeform*, leaving too many choices up to developers.

Some freedom *from choice* would be good, to reduce the variability, and make something like licenses much easier to analyse by tools.

Reply:
I understand the many good arguments for a freedom-from-choice solution but fear that in rare cases forcing SPDX may be a showstopper for integrating third-party SW (where a ROS developer has little influence on the license). There are examples of important libraries that come with common licenses plus some small additional clauses. This will likely cause headaches for someone who wants to use such a ROS package in a product. For the task of generating a machine-readable license/copyright list it is unproblematic, given that the corresponding source files and the license text are specified/linked properly in the `package.xml`.

Therefore, I prefer to stick with the approach of the [Machine-readable debian/copyright file]( https://www.debian.org/doc/packaging-manuals/copyright-format/) (as in many other aspects of this PR) which supports commonly used short names, freeform, and SPDX - and I like @clalancette’s idea of a `type` attribute, which can be used by linters as a safety layer for developers who want to follow the SPDX standard.

Reply:
+1 for the `type` attribute.

Reply:
Just for completeness: The SPDX Specification also describes a solution for naming custom licenses (cf. [https://spdx.dev/spdx-specification-21-web-version/#h.1v1yuxt](https://spdx.dev/spdx-specification-21-web-version/#h.1v1yuxt)), but I still prefer the explicit `type` attribute to distinguish between licenses from the [SPDX License List](https://spdx.org/licenses/) and custom ones.

Reply:
Introduced the `type` attribute as proposed by @clalancette. (Sorry, I should have updated the PR before yesterday's discussion in the TSC meeting.)

Is the description of the new attribute sufficient or should I give explicit explanations of the two possible values `freeform` and `spdx`?

Reply:
@clalancette, are you fine if I mark the discussion on the type attribute as resolved, cf. my last question on the explanation of the attribute in the REP?